### PR TITLE
CONSOLE-2321: Allow operators installed globally to display operand instances for all managed namespaces in their details

### PR DESCRIPTION
**Pull request by cyril-ui-developer**
/assign @spadgett 

The PR [New ListPage components](https://github.com/openshift/console/pull/8827) looks promising for adding widget to ListPage, however I don't know when it would merge. For now, I have added a props to MultiListPage. 

Also, I'm not sure if the Alert is still needed, so I left it out. 

### Before:

Screen Shot 2021-05-14 at 8 51 26 AM
Screen Shot 2021-05-14 at 8 51 35 AM

### After:
Screen Shot 2022-03-04 at 11 11 37 AM
Screen Shot 2022-03-04 at 11 11 48 AM
Screen Shot 2022-03-04 at 11 12 00 AM
Screen Shot 2022-03-04 at 11 12 36 AM

### CODE REVIEW

**Comment by spadgett** [comment]
cc @alimobrem @tlwu2013

File: frontend/packages/operator-lifecycle-manager/src/components/operand/index.tsx
Comment (spadgett) [inline frontend/packages/operator-lifecycle-manager/src/components/operand/index.tsx:170]: This might not be new, but we'd need to pull this into the render function to translate the group label.

File: frontend/packages/operator-lifecycle-manager/src/components/operand/index.tsx
Comment (spadgett) [inline frontend/packages/operator-lifecycle-manager/src/components/operand/index.tsx:514]: Do we need to check either one or the other based on `showNs`?

File: frontend/packages/operator-lifecycle-manager/src/components/operand/index.tsx
Comment (spadgett) [inline frontend/packages/operator-lifecycle-manager/src/components/operand/index.tsx:576]: Doesn't `namespaced` depend on the kind?

```suggestion
    { kind: `${kindObj.apiGroup}~${kindObj.apiVersion}~${kindObj.kind}`, namespaced: kindObj.namespaced },
```

**Review by spadgett — COMMENTED** [review]
Thanks @cyril-ui-developer. It would be good for @TheRealJon to take a look as well since he knows this code better.

**Review by TheRealJon — COMMENTED** [review]
I have some comments on the code, but before I go through and review this, I'm wondering if it makes sense to just wait on this until the list page refactor merges. This will have to be refactored as part of that PR if it merges first which seems like is just creating unnecessary work. @spadgett wdyt?

File: frontend/packages/operator-lifecycle-manager/src/components/operand/index.tsx
Comment (dtaylor113) [inline frontend/packages/operator-lifecycle-manager/src/components/operand/index.tsx:150]: This return logic seems to indicate "does not have OtherManagedNamespaces", or "has one ns === ' ' " ?

**Comment by cyril-ui-developer** [comment]
@dtaylor113 This return logic means it has OtherManagedNamespaces.

**Comment by TheRealJon** [comment]
@spadgett I think we mentioned waiting on #8827,  but it seems like this is also dependent on #8949. 

Also, this story relates to #9142. We need to make sure the list of operands on the list page and the uninstall modal always match. They should definitely both be using the same API to get that list.

**Comment by eliranmal** [comment]
hey @cyril-ui-developer,

i have a note about the UI; in the filter component (*show operands in: &hellip;*), the radio buttons are not vertically aligned with the text.
is this caused by some interference on the page styles, or is it just a bug of the patternfly component(s) used?

**Comment by cyril-ui-developer** [comment]
> hey @cyril-ui-developer,
> 
> i have a note about the UI; in the filter component (_show operands in: …_), the radio buttons are not vertically aligned with the text. is this caused by some interference on the page styles, or is it just a bug of the patternfly component(s) used?

@eliranmal I think it is style issue. I will take a look at it.

File: frontend/packages/operator-lifecycle-manager/src/components/operand/index.tsx
Comment (TheRealJon) [inline frontend/packages/operator-lifecycle-manager/src/components/operand/index.tsx:372]: nit, this is already destructured
```suggestion
  const managesOtherNamespaces = hasOtherManagedNamespaces(olmObj);
```

File: frontend/packages/operator-lifecycle-manager/src/components/operand/index.tsx
Comment (TheRealJon) [inline frontend/packages/operator-lifecycle-manager/src/components/operand/index.tsx:407]: nit, already destructured
```suggestion
      Header={managesOtherNamespaces && toggleView ? AllNsHeader : CurrentNsHeader}
```

File: frontend/packages/operator-lifecycle-manager/src/components/operand/index.tsx
Comment (TheRealJon) [inline frontend/packages/operator-lifecycle-manager/src/components/operand/index.tsx:405]: I'd love to move the nested EmtpyMsg component out of this component as something like "DefaultEmptyMessage" so that we aren't declaring a component inside another component. We also probably need to accept a prop that can override the default empty message:
```suggestion
      EmptyMsg={props.EmptyMsg ?? DefaultEmptyMsg}
```

File: frontend/packages/operator-lifecycle-manager/src/components/operand/index.tsx
Comment (TheRealJon) [inline frontend/packages/operator-lifecycle-manager/src/components/operand/index.tsx:416]: Remove reference to legacy firehose name. "getK8sWatchResources" or something along those lines.

File: frontend/packages/console-dynamic-plugin-sdk/src/extensions/console-types.ts
Comment (rawagner) [inline frontend/packages/console-dynamic-plugin-sdk/src/extensions/console-types.ts:383]: this has to stay required as `filter` is the actual filtering function to be applied on the data

File: frontend/packages/operator-lifecycle-manager/src/components/operand/index.tsx
Comment (rawagner) [inline frontend/packages/operator-lifecycle-manager/src/components/operand/index.tsx:498]: the filter function
```suggestion
            items: Object.keys(watchedResources).map((kind) => ({
              id: kindForReference(kind),
              title: kindForReference(kind),
            })),
            filter: (filters, resource) => {
              if (!filters || !filters.selected || !filters.selected.length) {
                return true;
              }
              return filters.selected.includes(resource.kind);
            },
```

but please try it, I did not :)

File: frontend/packages/operator-lifecycle-manager/src/components/operand/index.tsx
Comment (rawagner) [inline frontend/packages/operator-lifecycle-manager/src/components/operand/index.tsx:488]:
```suggestion
  const rowFilters: RowFilter<K8sResourceCommon>[] =
```

**Review by rawagner — CHANGES_REQUESTED** [review]

**Review by TheRealJon — CHANGES_REQUESTED** [review]
I had a couple of smaller comments and one larger refactor that we should consider.

The `ToggleValue` component and `useToggleValuesChange` hook could be tightened up a bit. It would simplify things quite a bit if we just used a single boolean flag in redux state, and manage that state within the component that renders the radio group interaction. We could then separate the radio group state (which radio button is selected) from the redux flag that indicates whether or not we should render all operands. This would also simplify the hook we use to get/set redux state. Here's an untested example:

New hook in `frontend/packages/operator-lifecycle-manager/components/operand/useShowOperandsInAllNamespaces.ts`:
```ts
// This hook can be used to consume and update the showOperandsInAllNamespaces redux state
export const useShowOperandsInAllNamespaces = () => {
  const dispatch = useDispatch();

  // Retrieve curernt boolean flag value from redux
  const showOperandsInAllNamespaces = useSelector((state) =>
    state.UI.get('showOperandsInAllNamespaces'),
  );

  // Basic setter callback for the above redux state
  const setShowOperandsInAllNamespaces = React.useCallback(
    (value: boolean) => dispatch(uiActions.setShowOperandsInAllNamespaces(value)),
    [dispatch],
  );
  
  // Rerturn current state and setter
  return [showOperandsInAllNamespaces, setShowOperandsInAllNamespaces];
};
```

New component in `frontend/packages/operator-lifecycle-manager/components/operand/ShowOperandsInAllNamespacesRadioGroup.tsx`:
```tsx
// State management is now handled within this component and caller can just render it.
export const ShowOperandsInAllNamespacesRadioGroup = () => {
  const { t } = useTranslation();
  const [
    showOperandsInAllNamespaces,
    setShowOperandsInAllNamespaces,
  ] = useShowOperandsInAllNamespaces();
  return (
    <RadioGroup
      label={t('olm~Show operands in:')}
      currentValue={showOperandsInAllNamespaces ? 'true' : 'false'}
      inline
      items={[
        {
          value: 'true',
          title: t('olm~All namespaces'),
        },
        {
          value: 'false',
          title: t('olm~Current namespace only'),
        },
      ]}
      onChange={({ currentTarget }) =>
        setShowOperandsInAllNamespaces(currentTarget.value === 'true')
      }
    />
  );
};
```

Now, in the `ProvidedAPI(s)Page` components, we can just render the ShowOperandsFromAllNamespacesRadioGroup component:
```tsx
//...
<ListPageHeader title={showTitle ? t('olm~All Instances') : undefined}>
  <div className="co-operator-details__toggle-value">
    {allManagesNamespaces && (
      <ShowOperandsInAllNamespacesRadioGroup />
    )}
  </div>
  <ListPageCreateDropdown onClick={createNavigate} items={createItems}>
    {t('olm~Create new')}
  </ListPageCreateDropdown>
</ListPageHeader>;
```

And wherever we need to consume the `showOperandsInAllNamespaces` redux state:
```ts
const [showOperandsInAllNamespaces] = useShowOperandsInAllNamespaces();
```

Let me know if you have any questions or want to talk through this some more!

File: frontend/packages/operator-lifecycle-manager/src/components/operand/useShowOperandsInAllNamespaces.ts
Comment (TheRealJon) [inline frontend/packages/operator-lifecycle-manager/src/components/operand/useShowOperandsInAllNamespaces.ts:26]: Convention is to return a tuple from a hook:

```suggestion
type UseShowOperandsInAllNamespaces = () => [boolean, (value: boolean) => void];
// This hook can be used to consume and update the showOperandsInAllNamespaces redux state
export const useShowOperandsInAllNamespaces: UseShowOperandsInAllNamespaces = () => {
  const dispatch = useDispatch();

  const showOperandsInAllNamespaces = useSelector((state: RootState) =>
    state.UI.get('showOperandsInAllNamespaces'),
  );

  const setShowOperandsInAllNamespaces = React.useCallback(
    (value: boolean) => dispatch(UIActions.setShowOperandsInAllNamespaces(value)),
    [dispatch],
  );

  return [showOperandsInAllNamespaces, setShowOperandsInAllNamespaces];
};
```

**Review by TheRealJon — COMMENTED** [review]
Need to run `yarn i18n` and commit those changes, plus a couple of other tweaks.

**Review by TheRealJon — APPROVED** [review]
/lgtm

**Comment by jhadvig** [comment]
QE Approver: 
/assign @XiyunZhao
Docs Approver: 
/assign @opayne1
PX Approver:	 
/assign @RickJWagner

**Comment by XiyunZhao** [comment]
Hi @cyril-ui-developer, please help to review the test case that I shared on UserStory: CONSOL-2321
This PR has been tested on a private build, some issues need your confirmation, all expected result is based on UX design doc.
1. An Alert banner is missing below the radio button row when 'All namespaces' is selected
image
2. Error Loading/404 not found page will be shown after clicking "Current namespace only" for the operator like "Red Hat OpenShift Pipelines"
image
image
3. Namespace value is missing on the list when selecting "All namespaces" for operators like "Red Hat OpenShift Pipelines"
image
4. Sort function doesn't work on "Namespaces" column, this issue can use Operator "Red Hat Integration -Camel K" for checking
image

**Comment by cyril-ui-developer** [comment]
@XiyunZhao  I have reviewed and approved the test case. 
1. The Alert feature was not implemented because we thought the “All Namespaces” label was already displayed in the radio option. Although the change was not updated in the UX design cc: @spadgett 
You can open a Bugzilla for 2, 3 and 4 so I can investigate them.

**Comment by XiyunZhao** [comment]
@cyril-ui-developer Thanks for your response, Bug [2091542](https://bugzilla.redhat.com/show_bug.cgi?id=2091542), [2091545](https://bugzilla.redhat.com/show_bug.cgi?id=2091545), [2091555](https://bugzilla.redhat.com/show_bug.cgi?id=2091555) are ready for investigate.

**Comment by XiyunZhao** [comment]
The description for the feature has been updated on UserStory, no other issue found, approved

**Pull request merged**
